File: Wrapping/Spidermonkey/WrappingTest/src/Mix.cpp
#include "Mix.hpp"
#include "GlobalFunctions.hpp"

//---------------------------------------------------
// Builtin types: bool, integer, floating point
//---------------------------------------------------
namespace Builtin
{
	//---------------------------------------------------
	// VOID FUNCTION, NO PARAMETER
	//---------------------------------------------------

	void void_param0(){ lastFunc = "void_param0"; }

	//---------------------------------------------------
	// VOID FUNCTION, 1 PARAMETER
	//---------------------------------------------------

	void void_param1_bool            (bool x)            { lastFunc = "void_param1_bool"; lastParam1.str((x==true) ? "true" : "false"); }

	void void_param1_char            (char x)            { lastFunc = "void_param1_char"; lastParam1.str(""); lastParam1 << (int)x;}
	void void_param1_signed_char     (signed char x)     { lastFunc = "void_param1_signed_char"; lastParam1.str(""); lastParam1 << (int)x;}
	void void_param1_unsigned_char   (unsigned char x)   { lastFunc = "void_param1_unsigned_char"; lastParam1.str(""); lastParam1 << (int)x;}
	void void_param1_short           (short x)           { lastFunc = "void_param1_short"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_unsigned_short  (unsigned short x)  { lastFunc = "void_param1_unsigned_short"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_int             (int x)             { lastFunc = "void_param1_int"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_unsigned_int    (unsigned int x)    { lastFunc = "void_param1_unsigned_int"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_long            (long x)            { lastFunc = "void_param1_long"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_unsigned_long   (unsigned long x)   { lastFunc = "void_param1_unsigned_long"; lastParam1.str(""); lastParam1 << x;}

	void void_param1_float           (float x)           { lastFunc = "void_param1_float"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_double          (double x)          { lastFunc = "void_param1_double"; lastParam1.str(""); lastParam1 << x;}
	void void_param1_long_double     (long double x)     { lastFunc = "void_param1_long_double"; lastParam1.str(""); lastParam1 << x;}

	void void_param1_wchar_t         (wchar_t x)         { lastFunc = "void_param1_wchar_t"; lastParam1.str(""); lastParam1 << (int)x;}

	//---------------------------------------------------
	// RETURN VALUE, NO PARAMETER
	//---------------------------------------------------

	bool           bool_param0            (){ lastFunc = "bool_param0"; return true; }

	char           char_param0            (){ lastFunc = "char_param0"; return 8; }
	signed char    signed_char_param0     (){ lastFunc = "signed_char_param0"; return 8; }
	unsigned char  unsigned_char_param0   (){ lastFunc = "unsigned_char_param0"; return 8; }
	short          short_param0           (){ lastFunc = "short_param0"; return 8; }
	unsigned short unsigned_short_param0  (){ lastFunc = "unsigned_short_param0"; return 8; }
	int            int_param0             (){ lastFunc = "int_param0"; return 8; }
	unsigned int   unsigned_int_param0    (){ lastFunc = "unsigned_int_param0"; return 8; }
	long           long_param0            (){ lastFunc = "long_param0"; return 8; }
	unsigned long  unsigned_long_param0   (){ lastFunc = "unsigned_long_param0"; return 8; }

	float          float_param0           (){ lastFunc = "float_param0"; return 8.5f; }
	double         double_param0          (){ lastFunc = "double_param0"; return 8.5; }
	long double    long_double_param0     (){ lastFunc = "long_double_param0"; return 8.5; }

	wchar_t        wchar_t_param0         (){ lastFunc = "wchar_t_param0"; return 8; }

	//---------------------------------------------------
	// OTHER COMBINATIONS
	//---------------------------------------------------
	void void_param1_int_float  (int x, float y)  { lastFunc = "void_param1_int_float"; lastParam1.str(""); lastParam1 << x; lastParam2.str(""); lastParam2 << y;}
	int  int_param1_int         (int x)           { lastFunc = "int_param1_int";  lastParam1.str(""); lastParam1 << x; return 8; }
	int  int_param1_int_float   (int x, float y)  { lastFunc = "int_param1_int_float"; lastParam1.str(""); lastParam1 << x; lastParam2.str(""); lastParam2 << y; return 8; }
}

//---------------------------------------------------
// Strings: character arrays, std::string
//---------------------------------------------------
namespace Strings
{
	//---------------------------------------------------
	// VOID FUNCTION, 1 PARAMETER
	//---------------------------------------------------
	void void_param0_constCharPtr        (const char* str){lastFunc = "void_param0_constCharPtr"; lastParam1.str(""); lastParam1 << str;}

	void void_param0_stdString           (std::string str){lastFunc = "void_param0_stdString"; lastParam1.str(""); lastParam1 << str;}
	void void_param0_constStdStringRef   (const std::string& str){lastFunc = "void_param0_constStdStringRef"; lastParam1.str(""); lastParam1 << str;}

	// original JSStrings will not be effected
	void void_param0_charPtr             (char* str){lastFunc = "void_param0_charPtr"; lastParam1.str(""); lastParam1 << str;}
	void void_param0_stdStringRef        (std::string& val){lastFunc = "void_param0_stdStringRef"; lastParam1.str(""); lastParam1 << val;}

	//---------------------------------------------------
	// RETURN VALUE, NO PARAMETER
	//---------------------------------------------------

	std::string stdString("ReturnString");

	const char*          constCharPtr_param0       (){ lastFunc = "constCharPtr_param0"; return "ReturnString"; }

	std::string          stdString_param0          (){ lastFunc = "stdString_param0"; return stdString; }
	const std::string&   constStdStringRef_param0  (){ lastFunc = "constStdStringRef_param0"; return stdString; }
	const std::string*   constStdStringPtr_param0  (){ lastFunc = "constStdStringPtr_param0"; return &stdString; }

	// original C++ strings will not be effected
	std::string&         stdStringRef_param0       (){ lastFunc = "stdStringRef_param0"; return stdString; }
	std::string*         stdStringPtr_param0       (){ lastFunc = "stdStringPtr_param0"; return &stdString; }
}

//---------------------------------------------------
// A simple class with members
//---------------------------------------------------
namespace Classes
{
	int SampleClass::memberFunc(float param){ lastInstance = (int)this; lastFunc = "memberFunc"; lastParam1.str(""); lastParam1 << param;  return 8; }
	void SampleClass::staticMemberFunc(){ lastInstance = 0; lastFunc = "staticMemberFunc"; };
}

//---------------------------------------------------
// Strings: Inheritance
//---------------------------------------------------
namespace Inheritance
{
	void BaseClass::baseFunc(){ lastInstance = (int)this; lastFunc = "baseFunc"; };
	void SubClass::subFunc()  { lastInstance = (int)this; lastFunc = "subFunc"; };
}

//---------------------------------------------------
// Using classes and ownership
//---------------------------------------------------
namespace UsingAndOwnership
{
	void ClassScriptOwned::func(){ lastInstance = (int)this; lastFunc = "func"; };
	void passScriptOwnedCopy(ClassScriptOwned copy);
	void passScriptOwnedRef(ClassScriptOwned& ref);
	void passScriptOwnedPtr(ClassScriptOwned* ptr);
	ClassScriptOwned  returnScriptOwnedCopy(){return ClassScriptOwned(); }
	ClassScriptOwned& returnScriptOwnedRef(){ return *(new ClassScriptOwned()); }
	ClassScriptOwned* returnScriptOwnedPtr(){ return new ClassScriptOwned(); }
	
	void ClassNativeOwned::func(){ lastInstance = (int)this; lastFunc = "func"; };
	void passNativeOwnedCopy(ClassNativeOwned copy){}
	void passNativeOwnedRef(ClassNativeOwned& ref){}
	void passNativeOwnedPtr(ClassNativeOwned* ptr){}
	
	// should be prohibited by restrictions
	ClassNativeOwned  returnNativeOwnedCopy(){return ClassNativeOwned(); }
	
	ClassNativeOwned instance;
	ClassNativeOwned& returnNativeOwnedRef(){ return instance; }
	ClassNativeOwned* returnNativeOwnedPtr(){ return &instance; }
}

File: Wrapping/Spidermonkey/WrappingTest/jstests/Mix.js



print("INFO: Running various test cases")
print("INFO: If a test case fails, then an error message is printed. Otherwise the test succeeded.")
print("INFO: If an exception is shown, there was an error somewhere")

//---------------------------------------------------
// Builtin
//---------------------------------------------------
print("\nINFO: Testing builtin types")
if(checkPropertyOn("Builtin", this, "Global Object"))
{
	testFunctionsOn(Builtin, { 
			"void_param0": {},
			"void_param1_bool": {param1: true, param1Expected: "true"},
			"void_param1_char": {param1: 10, param1Expected: "10"},
			"void_param1_signed_char": {param1: 10, param1Expected: "10"},
			"void_param1_unsigned_char": {param1: 10, param1Expected: "10"},
			"void_param1_short": {param1: 10, param1Expected: "10"},
			"void_param1_unsigned_short": {param1: 10, param1Expected: "10"},
			"void_param1_int": {param1: 10, param1Expected: "10"},
			"void_param1_unsigned_int": {param1: 10, param1Expected: "10"},
			"void_param1_long": {param1: 10, param1Expected: "10"},
			"void_param1_unsigned_long": {param1: 10, param1Expected: "10"},
			"void_param1_float": {param1: 10.5, param1Expected: "10.5"},
			"void_param1_double": {param1: 10.5, param1Expected: "10.5"},
			"void_param1_long_double": {param1: 10.5, param1Expected: "10.5"},
			"void_param1_wchar_t": {param1: 10, param1Expected: "10"},
			
			"bool_param0": {resultExpected: true},
			"char_param0": {resultExpected: 8},
			"signed_char_param0": {resultExpected: 8},
			"unsigned_char_param0": {resultExpected: 8},
			"short_param0": {resultExpected: 8},
			"unsigned_short_param0": {resultExpected: 8},
			"int_param0": {resultExpected: 8},
			"unsigned_int_param0": {resultExpected: 8},
			"long_param0": {resultExpected: 8},
			"unsigned_long_param0": {resultExpected: 8},
			"float_param0": {resultExpected: 8.5},
			"double_param0": {resultExpected: 8.5},
			"long_double_param0": {resultExpected: 8.5},
			"wchar_t_param0": {resultExpected: 8},
			
			"void_param1_int_float": {param1: 10, param1Expected: "10", param2: 10.5, param2Expected: "10.5"},
			"int_param1_int": {resultExpected: 8, param1: 10, param1Expected: "10"},
			"int_param1_int_float": {resultExpected: 8,param1: 10, param1Expected: "10", param2: 10.5, param2Expected: "10.5"}
		});
}

//---------------------------------------------------
// Strings
//---------------------------------------------------
print("\nINFO: Testing Strings")
if(checkPropertyOn("Strings", this, "Global Object"))
{
	testFunctionsOn(Strings, { 
			"void_param0_constCharPtr": {param1: "StringParam", param1Expected: "StringParam"},
			"void_param0_stdString": {param1: "StringParam", param1Expected: "StringParam"},
			"void_param0_constStdStringRef": {param1: "StringParam", param1Expected: "StringParam"},
			"void_param0_charPtr": {param1: "StringParam", param1Expected: "StringParam"},
			"void_param0_stdStringRef": {param1: "StringParam", param1Expected: "StringParam"},

			"constCharPtr_param0": {resultExpected: "ReturnString"},
			"stdString_param0": {resultExpected: "ReturnString"},
			"constStdStringRef_param0": {resultExpected: "ReturnString"},
			"constStdStringPtr_param0": {resultExpected: "ReturnString"},
			"stdStringRef_param0": {resultExpected: "ReturnString"},
			"stdStringPtr_param0": {resultExpected: "ReturnString"},

		});
}

//---------------------------------------------------
// A simple class with members
//---------------------------------------------------
print("\nINFO: Testing SampleClass")
if(checkPropertyOn("Classes", this, "Global Object"))
{
	if(checkPropertyOn("SampleClass", Classes, "Classes"))
	{
		print("-- Testing static functions")
		testFunctionsOn(Classes.SampleClass, { "staticMemberFunc": {}});
	
		print("-- Testing instance")
		var instance = new Classes.SampleClass();
		if(!instance)
			print("ERROR: could not create instance of 'SampleClass'");
		else
		{
			testFunctionsOn(instance, { "memberFunc": { resultExpected: 8, param1: 8.5, param1Expected: "8.5" }});
			
			print("--- Testing calling member function with wrong execution context (exception expected)")
			try {
				instance.memberFunc.call({}, 8.5);
				print("---- ERROR: no exception thrown")
			} catch(e){
			}
		}
			
		
	}
}

//---------------------------------------------------
// ===== Helper functions for testing =====
//---------------------------------------------------
function testFunctionsOn(object, funcs)
{
	if(!object)
	{
		print("ERROR: Object'" + objectName + "' not existing");
		return false;
	}
	
	var fail = false;
	for(var funcName in funcs)
	{
		print("--- Testing " + funcName)
		if(funcName in object)
		{
			var result = object[funcName](funcs[funcName].param1, funcs[funcName].param2);
			verifyFunctionCall(funcName, result, funcs[funcName].resultExpected, funcs[funcName].param1Expected, funcs[funcName].param2Expected);
		}
		else
		{
			fail = true;
			print("---- ERROR: Function '" + funcName + "' is not defined on the given object");
		}
	}
	
	return fail;
}

function checkPropertyOn(propName, object, objectName)
{
	if(!object)
	{
		print("ERROR: Object'" + objectName + "' not existing");
		return false;
	}
	
	if(propName in object)
		return true;
	
	print("ERROR: '" + propName + "' not defined on '" + objectName + "'");
	return false;
}

function verifyFunctionCall(funcName, returnVal, returnValExpected, param1, param2, instance)
{
	if(getLastFunction() !== funcName)
		print("ERROR: function not called");
	
	if(typeof param1 !== "undefined" && getLastParam1() !== param1)
		print("ERROR: function converted parameter 1 wrong");
		
	if(typeof param2 !== "undefined" && getLastParam2() !== param2)
		print("ERROR: function converted parameter 2 wrong");
	
	if(typeof returnValExpected !== "undefined" && returnVal !== returnValExpected)
		print("ERROR: function returned wrong value");
}